### ISSUE TITLE
Root-implicit optimizer silently ignores gs_line_search, gs_optimizer='cg', gs_c4v and return_history (4 unresolved #773 comments, all still live)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#792]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Four review comments on #773 were never resolved. I re-verified all four against `main` today; **all four are still live**. Grouping them because they share one shape: `optimize_gs_ad_root_implicit` accepts a documented knob and then ignores it, so the caller gets a materially different optimizer with no warning.

### 1. `gs_line_search*` is silently ignored (P1)

`_use_line_search()` returns `True` by default for `lbfgs` (`ipeps_optimize.py:476-480`), but the root-implicit loop never calls a line search. `_build_optimizer` for `lbfgs` returns

```python
optax.chain(optax.scale_by_lbfgs(memory_size=10), optax.clip_by_global_norm(...), optax.scale(-1.0))
```

— no line-search transform. Passing `value_fn=` to `.update()` on that chain does nothing; only `optax.lbfgs(linesearch=...)` consumes it. So the **default** root-implicit optimizer takes unconditional quasi-Newton steps and can accept energy-increasing ones, while `gs_line_search`, `gs_line_search_method`, `gs_line_search_max_steps` and `gs_hz_max_iter` are all no-ops.

The in-code comment ("optax's L-BFGS line search re-evaluates the objective... Pass only the value function") reads as though a line search is running. It is not.

### 2. `gs_optimizer="cg"` is plain gradient descent (P2)

`_build_optimizer` returns `None` for `cg`, and the loop then does

```python
params = _normalize_params(params - config.gs_learning_rate * grads)
```

No previous gradient retained, no Polak-Ribière coefficient, no conjugate direction, and no line search — which defaults *on* for CG.

### 3. `gs_c4v=True` is accepted but not enforced (P2)

No `c4v` handling anywhere in `ipeps_optimize_root_implicit.py`. The dispatcher accepts the combination for a 1x1 run, but the path only normalises the full asymmetric tensor — it never projects into the C4v basis or optimises reduced coefficients, so the first update can leave the requested symmetry sector.

### 4. `return_history=True` silently drops the history (P2)

No `return_history` handling on this path. The unconditional three-element return means callers of the documented 1x1 history API get the wrong tuple shape and lose energies/timings.

### Suggested resolution

For each: either implement it, or **reject it explicitly at the dispatcher**. Silently accepting is the worst of the three, and #755 is already open for the same class of defect (`gs_recipe` ignored on the explicit-AD path) — worth fixing together with a single "knobs this path cannot honour" gate, which `ipeps_optimize_root_implicit` already has the beginnings of (`test_the_error_names_every_offending_knob_at_once`).

Refs #773, #715. Related: #755.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Fixed by PR #811, merged as `74fa8f3`. (The PR said "Refs", not "Closes", so this needed closing by hand.)

All four knobs were re-verified live before the fix, and they did **not** all get the same treatment — that distinction is the substance of what shipped:

| knob | default | treatment | why |
|---|---|---|---|
| `gs_optimizer='cg'` | opt-in | **raise** | the loop ran plain gradient descent under the CG name |
| `gs_c4v=True` | opt-in | **raise** | nothing projects into the C4v basis |
| `gs_line_search` | **auto-on** | **warn** | see below |
| `return_history` | opt-in | **implement** | documented API; the loop already had the data |

**The suggested resolution here — reject all four at the dispatcher — would have been wrong for `gs_line_search`.** `gs_optimizer` defaults to `lbfgs`, and `gs_line_search=None` means auto-on for lbfgs/cg (`_use_line_search`), so a blanket rejection raises on the path's own **default** configuration and on the documented Path 5 example. It warns instead, following the rule this file already set for `gs_metric_precond`: **refuse what was explicitly requested, warn about what arrived by default.**

Two further defects were caught in review and fixed in the same PR:

1. **`return_history` violated the shape invariant** the other AD paths are held to (`tests/test_ipeps_ad_history.py:62`, `len(step_times) == max(num_steps - 1, 0)`). Every step including the compile-dominated first evaluation was appended to `step_times` while `jit_compile_time` was hard-coded `0.0`. This path has no top-level `jax.jit`, but its root solve and adjoint are jitted internally, so step 0 still pays their lazy compile — the benchmark two-point compile/steady-state fit would have taken a cold sample as warm.
2. **`converged=True` could be published for a run whose gradient went non-finite.** Masking those entries to zero drives `grad_norm` to exactly `0.0`, firing `_converged_outer` on a step this loop already warns is "a no-op and apparent convergence here is not convergence". The flag is now gated on `nonfinite_grad_steps == 0`.

Verified on `main` (`74fa8f3`): knob gates, the conditional warning, `first_step`/`jit_compile_time` and `converged_flag = nonfinite_grad_steps == 0` all present. CI green on all required checks; independently green on CPU (39 passed) and GPU (35 passed). Mutation-verified — each mutation kills exactly its own test.

**Not fixed here, deliberately:** the loop still *breaks* on that fake convergence. Only the reported flag was corrected. The break is pre-existing, affects every root-implicit caller rather than just `return_history` ones, and shifts returned energies and step counts — tracked as **#812**.

Refs #773, #715.
